src/core/cmcd/cmcd_data_builder.ts

```diff
@@ -0,0 +1,602 @@
+            props.bl = Math.floor(Math.round(precizeBufferLengthMs / 100) * 100);
+            break;
+          }
+        }
+      }
+    }
+
+    const precizeDeadlineMs =
+      precizeBufferLengthMs === undefined || lastObservation === undefined
+        ? undefined
+        : precizeBufferLengthMs / lastObservation.speed;
+
+    props.dl =
+      precizeDeadlineMs === undefined
+        ? undefined
+        : Math.floor(Math.round(precizeDeadlineMs / 100) * 100);
+
+    if (precizeDeadlineMs !== undefined) {
+      // estimate the file size, in kilobits
+      const estimatedFileSizeKb =
+        (content.representation.bitrate * content.segment.duration) / 1000;
+      const wantedCeilBandwidthKbps = estimatedFileSizeKb / (precizeDeadlineMs / 1000);
+      props.rtp = Math.floor(
+        Math.round((wantedCeilBandwidthKbps * RTP_FACTOR) / 100) * 100,
+      );
+    }
+
+    switch (content.manifest.transport) {
+      case "dash":
+        props.sf = "d";
+        break;
+      case "smooth":
+        props.sf = "s";
+        break;
+      default:
+        props.sf = "o";
+        break;
+    }
+    props.st = content.manifest.isDynamic ? "l" : "v";
+    props.tb = content.adaptation.representations.reduce(
+      (acc: number | undefined, representation: IRepresentation) => {
+        if (
+          representation.isSupported !== true ||
+          representation.decipherable === false
+        ) {
+          return acc;
+        }
+        if (acc === undefined) {
+          return Math.round(representation.bitrate / 1000);
+        }
+        return Math.max(acc, Math.round(representation.bitrate / 1000));
+      },
+      undefined,
+    );
+    return this._producePayload(props);
+  }
+
+  private _producePayload(props: ICmcdProperties): ICmcdPayload {
+    let cmcdObjectValue = "";
+    let cmcdRequestValue = "";
+    let cmcdSessionValue = "";
+    let cmcdStatusValue = "";
+    let queryStringPayload = "";
+    if (props.br !== undefined) {
+      queryStringPayload += `br=${String(props.br)},`;
+      const toAdd = `br=${String(props.br)},`;
+      if (this._typePreference === TypePreference.Headers) {
+        cmcdObjectValue += toAdd;
+      } else {
+        queryStringPayload += toAdd;
+      }
+    }
+    if (props.bl !== undefined) {
+      const toAdd = `bl=${String(props.bl)},`;
+      if (this._typePreference === TypePreference.Headers) {
+        cmcdRequestValue += toAdd;
+      } else {
+        queryStringPayload += toAdd;
+      }
+    }
+    if (props.bs === true) {
+      const toAdd = "bs,";
+      if (this._typePreference === TypePreference.Headers) {
+        cmcdStatusValue += toAdd;
+      } else {
+        queryStringPayload += toAdd;
+      }
+    }
+    if (props.cid !== undefined) {
+      const toAdd = `cid=${formatStringPayload(props.cid)},`;
+      if (this._typePreference === TypePreference.Headers) {
+        cmcdSessionValue += toAdd;
+      } else {
+        queryStringPayload += toAdd;
+      }
```

Comment:
there is probably a way to factorize this

Reply:
Yeah, the shaka-something did so: 

https://github.com/shaka-project/shaka-player/blob/88d2a0227bc23ae2e02d45fd3fc27cd3d3e3f66d/lib/util/cmcd_manager.js#L699

and

https://github.com/shaka-project/shaka-player/blob/88d2a0227bc23ae2e02d45fd3fc27cd3d3e3f66d/lib/util/cmcd_manager.js#L771

But I usually prefer easy-to-read dumb code to shorter big-brain code.

Though you're right and due to that repetition the issue like seen for `br` are more probable and worse, easy to miss.
Maybe we can find some middle way?

Reply:
Updated

Reply:
Nice